src/SMatrix.jl

```diff
@@ -123,6 +123,7 @@ end
     Base.unsafe_convert(Ptr{T}, Base.data_pointer_from_objref(m))
 end
 
+Base.promote_rule{S1,S2,T,U,L}(::Type{SMatrix{S1,S2,T,L}}, ::Type{SMatrix{S1,S2,U,L}}) = SMatrix{S1,S2,promote_type(T,U),L}
```

Comment:
Does it do anything? The compiler treats type transformations as pure since the types come out of type inference anyway. (i.e. we get `Type{SMatrix{S1,S2...}}` out of this which is *not* widened in the compiler to `::DataType`)

If `promote_type(T,U)` was sufficiently complex (uninferable) perhaps `@pure` would be useful, but I think its up to the devs of `T` and `U` to decide if they make their `promote_rule` pure.